example/src/main/scala/example/domain/Contributors.scala

```diff
@@ -0,0 +1,55 @@
+package example.domain
+
+import io.circe
+import io.circe.syntax._
+import zio._
+import zio.redis._
+import io.circe.parser.decode
+import sttp.client.asynchttpclient.zio.SttpClient
+import sttp.client.{UriContext, basicRequest}
+import sttp.client.circe.asJson
+import sttp.model.Uri
+import Contributor._
+import zio.duration._
+
+object Contributors {
+
+  trait Service {
+    def getContributors(organization: String, repository: String): IO[ApiError, Chunk[Contributor]]
+  }
+
+  lazy val live: ZLayer[RedisExecutor with SttpClient, Nothing, Contributors] =
+    ZLayer.fromFunction { env =>
+      new Service {
+        override def getContributors(organization: String, repository: String): IO[ApiError, Chunk[Contributor]] =
+          sMembers(repository).flatMap { response => // key should be unique combination of org + repo
+              if (response.isEmpty)
+                for {
+                  contributors <- fetchContributors(organization, repository)
+                  _ <- sAdd(repository, contributors.asJson.toString, contributors.map(_.asJson.toString):_*)
```

Comment:
Well, this isn't correct. First of all you're saving the stuff twice. Second, `contributors` might be empty. That said, i propose the following:

- Build a non-empty chunk for "cacheable" data, or return immediately if no contributors are found.
- Once you built a non-empty chunk, calculate `serialized = contributors.map(_.asJson.noSpaces)`
- Invoke `SADD` as `sAdd(key, serialized.head, serialized.tail: _*)`

Once this is working, we can reintroduce your `Cache` abstraction as:

```scala
object Cache {
  trait Service {
    // you can specialize
    def readThrough[K, E, A](key: K, lookup: => IO[E, A]): IO[E, A]
  }
}
```